infer/src/IR/Attributes.ml

```diff
@@ -94,6 +94,55 @@ let load, clear_cache, store =
   in
   (load, clear_cache, store)
 
+let update_source_file =
+  let select_existing_statement =
+    ResultsDatabase.register_statement
+      "SELECT procedure_names FROM source_files WHERE source_file = :source"
+  in
+  let load_proc_names source_file =
+    ResultsDatabase.with_registered_statement select_existing_statement ~f:(fun db stmt ->
+      SourceFile.SQLite.serialize source_file
+      |> Sqlite3.bind stmt 1
+      (* :source *)
+      |> SqliteUtils.check_result_code db ~log:"get_existing_data bind source file" ;
+      SqliteUtils.result_option ~finalize:false db ~log:"looking for pre-existing source file data"
+        stmt ~read_row:(fun stmt ->
+          Sqlite3.column stmt 0 |> Procname.SQLiteList.deserialize ) )
+  in
+  fun ~source_file ~tenv ~integer_type_widths ~procname ->
+    let new_proc_names = 
+      match load_proc_names source_file with
+      | Some old_proc_names ->
+        if not (List.mem old_proc_names procname ~equal:Procname.equal) then 
+          List.append old_proc_names [procname]
+        else
+          old_proc_names
+      | None ->
+        [procname]
+    in
+    DBWriter.add_source_file
+      ~source_file:(SourceFile.SQLite.serialize source_file) 
+      ~tenv:(Tenv.SQLite.serialize tenv) 
+      ~integer_type_widths:(Typ.IntegerWidths.SQLite.serialize integer_type_widths)
+      ~proc_names:(Procname.SQLiteList.serialize new_proc_names)
+
+let store ~proc_desc (attr : ProcAttributes.t) =
+  let pname = attr.proc_name in
+  let akind = proc_kind_of_attr attr in
+  let proc_uid = Procname.to_unique_id pname in
+  if should_try_to_update proc_uid akind then
+    let proc_name = Procname.SQLite.serialize pname in
+    let attr_kind = int64_of_attributes_kind akind in
+    let cfg = Procdesc.SQLite.serialize proc_desc in
+    let proc_attributes = ProcAttributes.SQLite.serialize attr in
+    let source_file = SourceFile.SQLite.serialize attr.loc.Location.file in
+    let callees =
+      Option.value_map proc_desc ~f:Procdesc.get_static_callees ~default:[]
+      |> Procname.SQLiteList.serialize
+    in
+    DBWriter.replace_attributes ~proc_uid ~proc_name ~attr_kind ~source_file ~proc_attributes ~cfg ~callees ;
```

Comment:
This bypasses and invalidates the caching done above. That logic should be moved in the main `store` function above instead of in a copy of `store` without the cache.

Can you say more about why the changes in this file are needed? Why do you need to update things in the DB after the fact?

Reply:
Sorry it seems that I have created a duplicate `store` function. Now it has been removed and fixed.

We need to update DB when new procedure names were added in changed files. So Infer can be notified to analyze these added procedures as well